lib/elixir/lib/enum.ex

```diff
@@ -723,28 +723,30 @@ defmodule Enum do
   end
 
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
-      {:error, _module} ->
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
```

Comment:
@michalmuskala we do a double dispatch on the worst case scenario only. You do a double dispatch on all cases. If an implementation doesn't want to perform a double dispatch, then they can always implement `count`.